DMTN-148.tex

```diff
@@ -50,9 +50,276 @@
 % ADD CONTENT HERE
 % You can also use the \input command to include several content files.
 
+\section{Intro}
+
+The goal of this tech note is to provide a plan for the handling of
+calibration products in the Gen-3 Butler environment.  There are a
+large number of these products currently in use, but many of them do
+not have a clearly defined process to define, ingest, test, and
+deprecate them.  The lack of documentation on how these processes work
+has led to many implementations of similar paths, further complicating
+the issue.
+
+
+\section{The End State}
+
+The final state that we wish to get to has the following properties:
+
+\begin{itemize}
+\item All calibration products are either lsst.afw.image.Exposure
+  image data or an explicit calibration class in the \verb|ip_isr|
+  package.\footnote{This will require class reorganization to migrate
+    the Defects and Curve calibrations definitions.}
+\item All calibration products should be constructed from a DM Task in
+  the \verb|cp_pipe| package, and have the input data and
+  configuration parameters persisted with the calibration itself.
+  This guarantees that any product can be reconstructed in the future.
+  Exceptions will exist for external or lab created calibrations,
+  although the majority of used calibrations will likely be produced
+  directly.
```

Comment:
In the cases I can think of, the "input data" described here is a list of raw images; to the extend that the production process involves multiple `PipelineTasks`, we're generally assuming all steps are run together, and we're interested in their collective ultimate inputs and not recording intermediate datasets.  I'm fine with that - it duplicates some provenance information that will be recorded by the middleware system for _all_ `PipelineTasks`, but it serves as a simpler, more human-digestable summary that in at least some (probably most) cases is sufficient for full reproducibility, and unlike the middleware provenance it's something we can write directly to the files.  Those seem like good enough virtues to outweigh concerns about having two provenance systems.

But I'd be very wary of extending this to cases where the inputs being recorded are not raws, or any other scenario where we start to talk about _recursive_ provenance recording (i.e. one Task's inputs are another Task's outputs, and those aren't always produced at the same time or with consistent software versions).  That's a lot more complex - the fact that we can easily identify a raw uniquely with a single number across all data repositories is quite specia, and it's dangerous to assume something that works for raws can easily be extended to other kinds of datastes.  The middleware system is being designed to solve this more general problem, and we should not attempt to put together another implementation of something that complex.

If we ever need to store _recursive_ provenance in an `obs_CAMERA_data` package at any point, we should approach that from the perspective of exporting it from the data repository, not utilizing the parallel provenance-recording mechanism described here.

Reply:
I'm leaning more and more that this is a thing generated similar to the calibration's export.  This uses the middleware solution, and can be used to dump a snapshot of an entire calibration collection, which seems like the most likely use case.